Rubberduck.VBEEditor/ComManagement/TypeLibs/AddressableVariables.cs

```diff
@@ -0,0 +1,231 @@
+        public TMarshalled Value
+        {
+            get => GetArrayElement(0);
+            set => SetArrayElement(0, value);
+        }
+
+        // the individual size of each element
+        public int ElementSize => Marshal.SizeOf(typeof(TUnmarshalled));
+
+        // array accessors, unmarshalled
+        public TUnmarshalled GetArrayElementUnmarshalled(int elementIndex)
+        {
+            var elementAddress = Address + (ElementSize * elementIndex);
+            return StructHelper.ReadStructureUnsafe<TUnmarshalled>(elementAddress);
+        }
+        public void SetArrayElementUnmarshalled(int elementIndex, TUnmarshalled value)
+        {
+            var elementAddress = Address + (ElementSize * elementIndex);
+            Marshal.StructureToPtr<TUnmarshalled>(value, elementAddress, false);
+        }
+
+        // array accessors, marshalled
+        public TMarshalled GetArrayElement(int elementIndex)
+            => MarshalFrom(GetArrayElementUnmarshalled(elementIndex));
+
+        public void SetArrayElement(int elementIndex, TMarshalled value)
+            => SetArrayElementUnmarshalled(elementIndex, MarshalTo(value));
+
+        // GetArray: grab a full copy of the array content as a c# array, including marshalling of all elements
+        public TMarshalled[] GetArray(int maxCopyElements = 0)
+        {
+            if (maxCopyElements > ElementCount) throw new InvalidOperationException();
+            if (maxCopyElements == 0) maxCopyElements = ElementCount;
+
+            TMarshalled[] retVal = new TMarshalled[maxCopyElements];
+            var index = 0;
+            while (index < maxCopyElements)
+            {
+                retVal[index] = GetArrayElement(index);
+                index++;
+            }
+            return retVal;
+        }
+
+        // CopyArrayTo: copy the internal contiguous array to a c# array, including marshalling of all elements
+        // this assumes input array is of the correct size
+        public void CopyArrayTo(TMarshalled[] copyTo)
+        {
+            var index = 0;
+            while (index < ElementCount)
+            {
+                copyTo[index] = GetArrayElement(index);
+                index++;
+            }
+        }
+
+        public IAddressableVariableBase(int contiguousArrayElementCount = 1, IntPtr alreadyAllocatedMem = default)
+        {
+            ElementCount = contiguousArrayElementCount;
+            var sizeOf = ElementSize * ElementCount;
+            _ownedMemory = alreadyAllocatedMem == IntPtr.Zero;
+            if (_ownedMemory)
+            {
+                Address = Marshal.AllocHGlobal(sizeOf);
+                Marshal.Copy(new byte[sizeOf], 0, Address, sizeOf); // nullify the data
+            }
+            else
+            {
+                Address = alreadyAllocatedMem;
+            }
+        }
+
+        // Extract the address and ensure no cleanup
+        public IntPtr Extract()
```

Comment:
FWIW, it's not clear to me why the act of extracting would mean no cleanup must be done. Is it implied that if I `Extract`, I've taken the responsibility of cleaning it up? If that's the case, that should be made explicit.

Reply:
yeah, i'll sort that.   Not currently using that method, but in my head  I'm planning to